Presentation/CustomAdaptivePresentationDelegate.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  CustomAdaptivePresentationDelegate.swift
+//  Presentation
+//
+//  Created by Vasil Blanco-Nunev on 2019-07-19.
+//  Copyright © 2019 iZettle. All rights reserved.
+//
+
+import UIKit
+import Flow
+
+/// Exposes a reactive interface for a delegate conforming to `UIAdaptivePresentationControllerDelegate`
+public final class CustomAdaptivePresentationDelegate: NSObject, UIAdaptivePresentationControllerDelegate {
+    public var adaptivePresentationStyle = Delegate<(UIPresentationController, UITraitCollection?), UIModalPresentationStyle>()
+    public var viewControllerForAdaptivePresentationStyle = Delegate<(UIPresentationController, UIModalPresentationStyle), UIViewController?>()
+    public var shouldDismiss = Delegate<UIPresentationController, Bool>()
+
+    private let willPresentCallbacker = Callbacker<WillPresentAdaptivelyInput>()
+    private let willDismissCallbacker = Callbacker<UIPresentationController>()
+    private let didAttemptToDismissCallbacker = Callbacker<UIPresentationController>()
+    private let didDismissCallbacker = Callbacker<UIPresentationController>()
+
+    public typealias WillPresentAdaptivelyInput = (UIPresentationController, UIModalPresentationStyle, UIViewControllerTransitionCoordinator?)
+
+    public var willPresentSignal: Signal<WillPresentAdaptivelyInput> {
+        return Signal(callbacker: willPresentCallbacker)
+    }
+
+    /// A signal that fires on iOS 13+ when a modal presentation will get dismissed by swiping down.
+    ///
+    /// - Note: It fires when the dismissed view controller has `isModalInPresentation` set to `false`
+    public var willDismissSignal: Signal<UIPresentationController> {
+        return Signal(callbacker: willDismissCallbacker)
+    }
+
+    /// A signal that fires on iOS 13+  when a modal presentation attemts to get dismissed by swiping down
+    ///
+    /// - Note: For this to get called, the dismissed view controller needs to have `isModalInPresentation` set to `true`
+    public var didAttemptToDismissSignal: Signal<UIPresentationController> {
+        return Signal(callbacker: didAttemptToDismissCallbacker)
+    }
+
+    /// A signal that fires on iOS 13+  when a modal presentation gets dismissed by swiping down
+    public var didDismissSignal: Signal<UIPresentationController> {
+        return Signal(callbacker: didDismissCallbacker)
+    }
+
```

Comment:
Should these three signals be exposed for < iOS 13 users? Will they ever signal?

Reply:
I think that's nicer than adding iOS 13 check everywhere they are used. I specified it in the docs.